test/server/connectors/serviceInsights/graphDataExtractor.spec.js

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2019 Expedia Group
+ *
+ *       Licensed under the Apache License, Version 2.0 (the License);
+ *       you may not use this file except in compliance with the License.
+ *       You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *       Unless required by applicable law or agreed to in writing, software
+ *       distributed under the License is distributed on an AS IS BASIS,
+ *       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *       See the License for the specific language governing permissions and
+ *       limitations under the License.
+ *
+ */
+
+import {expect} from 'chai';
+import {extractNodesAndLinks} from '../../../../server/connectors/serviceInsights/graphDataExtractor';
+
+/* eslint-disable no-unused-expressions */
+describe('graphDataExtractor.extractNodesAndLinks', () => {
+    function randomId() {
+        return Math.floor(Math.random() * 10000);
+    }
+
+    function trace(...spans) {
+        const traceId = randomId();
+
+        spans.forEach((s, i) => {
+            s.traceId = traceId;
+            if (i > 0) {
+                s.parentSpanId = spans[i - 1].spanId;
+            }
+        });
+
+        return spans;
+    }
+
+    function span(serviceName, operationName, tags = []) {
+        const spanId = randomId();
+        return {
+            spanId,
+            serviceName,
+            operationName,
+            tags
+        };
+    }
+
+    /*
+     * The following span functions must match the server/config/base.js configuration for spanTypes.
+     */
+
+    function edgeSpan() {
```

Comment:
ps we had some other thoughts in the past such as synthetic query-time tags. For example extract a transformation function that backfills data like this in the span so that all screens can use it.